atuin-client/src/import/mod.rs

```diff
@@ -1,27 +1,46 @@
-use std::{
-    io::{BufRead, BufReader, Read, Seek, SeekFrom},
-    path::{Path, PathBuf},
-};
+use std::io;
 
+use async_trait::async_trait;
 use eyre::Result;
 
 use crate::history::History;
 
 pub mod bash;
-pub mod fish;
-pub mod resh;
-pub mod zsh;
+// pub mod fish;
+// pub mod resh;
+// pub mod zsh;
 
-// this could probably be sped up
-fn count_lines(buf: &mut BufReader<impl Read + Seek>) -> Result<usize> {
-    let lines = buf.lines().count();
-    buf.seek(SeekFrom::Start(0))?;
-
-    Ok(lines)
+fn count_lines(input: &[u8]) -> usize {
+    memchr::memchr_iter(b'\n', input).count()
 }
 
-pub trait Importer: IntoIterator<Item = Result<History>> + Sized {
+#[async_trait]
+pub trait Importer: Sized {
     const NAME: &'static str;
-    fn histpath() -> Result<PathBuf>;
-    fn parse(path: impl AsRef<Path>) -> Result<Self>;
+    async fn new() -> io::Result<Self>;
+    async fn entries(&mut self) -> Result<usize>;
+    async fn load(self, loader: &mut impl Loader) -> Result<()>;
+}
+
+#[async_trait]
+pub trait Loader: Sync + Send {
+    async fn push(&mut self, hist: History) -> eyre::Result<()>;
+}
```

Comment:
This is the core API change. It's very simple, the importer can

1. Be created
2. Count the number of entries it may contain
3. Load all of it's entries into a `Loader`. (A `TestLoader` is defined below for testing)

All these methods are async and fallible so I'm not sure how much more flexible we can be :sweat_smile: